go-controller/pkg/ovn/policy_test.go

```diff
@@ -1334,6 +1334,146 @@ var _ = ginkgo.Describe("OVN NetworkPolicy Operations", func() {
+					[]knet.NetworkPolicyEgressRule{{
+						Ports: []knet.NetworkPolicyPort{{
+							Port:     &intstr.IntOrString{IntVal: portNum + 1},
+							Protocol: &tcpProtocol,
+						}},
+					}},
+				)
+
+				npTest := kNetworkPolicy{}
+				gressPolicy1ExpectedData := npTest.getPolicyData(networkPolicy, []string{nPodTest.portUUID}, nil, []int32{portNum}, nbdb.ACLSeverityInfo)
+				defaultDenyExpectedData := npTest.getDefaultDenyData(networkPolicy, []string{nPodTest.portUUID}, nbdb.ACLSeverityInfo)
+				expectedData := []libovsdb.TestData{}
+				expectedData = append(expectedData, gressPolicy1ExpectedData...)
+				expectedData = append(expectedData, defaultDenyExpectedData...)
+				expectedData = append(expectedData, getExpectedDataPodsAndSwitches([]testPod{nPodTest}, []string{"node1"})...)
+
+				fakeOvn.startWithDBSetup(initialDB,
+					&v1.NamespaceList{
+						Items: []v1.Namespace{namespace1},
+					},
+					&v1.PodList{
+						Items: []v1.Pod{*nPod},
+					},
+					&knet.NetworkPolicyList{
+						Items: []knet.NetworkPolicy{*networkPolicy},
+					},
+				)
+				nPodTest.populateLogicalSwitchCache(fakeOvn, getLogicalSwitchUUID(fakeOvn.controller.nbClient, "node1"))
+				fakeOvn.controller.WatchNamespaces()
+				fakeOvn.controller.WatchPods()
+				fakeOvn.controller.WatchNetworkPolicy()
+
+				ginkgo.By("Creating a network policy that applies to a pod")
+
+				_, err := fakeOvn.fakeClient.KubeClient.NetworkingV1().NetworkPolicies(networkPolicy.Namespace).Get(context.TODO(), networkPolicy.Name, metav1.GetOptions{})
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				fakeOvn.asf.ExpectAddressSetWithIPs(namespaceName1, []string{nPodTest.podIP})
+				gomega.Eventually(fakeOvn.nbClient).Should(libovsdb.HaveData(expectedData...))
+
+				ginkgo.By("Bringing down NBDB")
+				// inject transient problem, nbdb is down
+				fakeOvn.controller.nbClient.Close()
+				gomega.Eventually(fakeOvn.controller.nbClient.Connected()).Should(gomega.BeFalse())
+
+				ginkgo.By("Delete the first network policy")
+				err = fakeOvn.fakeClient.KubeClient.NetworkingV1().NetworkPolicies(networkPolicy.Namespace).Delete(context.TODO(), networkPolicy.Name, metav1.DeleteOptions{})
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+				// sleep long enough for TransactWithRetry to fail, causing NP Add to fail
+				time.Sleep(types.OVSDBTimeout + time.Second)
+				// check to see if the retry cache has an entry for this policy
+				gomega.Eventually(fakeOvn.controller.getRetryEntry(networkPolicy2)).ShouldNot(gomega.BeNil())
+				retryEntry := fakeOvn.controller.getRetryEntry(networkPolicy2)
+				ginkgo.By("retry entry new policy should be nil")
+				gomega.Expect(retryEntry.newPolicy).To(gomega.BeNil())
+				ginkgo.By("retry entry old policy should not be nil")
+				gomega.Expect(retryEntry.oldPolicy).NotTo(gomega.BeNil())
+				connCtx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
+				defer cancel()
+				fakeOvn.resetNBClient(connCtx)
```

Comment:
@dave-tucker @jcaamano there appears to be some bug in libovsdb here. After resetNBClient, which looks like this:
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ovn_test.go#L110

Sometime later the client will complain:
```
2022/02/15 18:01:28 server/cache: "caller"={"file":"cache.go","line":650} "level"=5 "msg"="processing update" "uuid"="a6114f2c-c662-4c1e-bdf4-cc83b3ffaf3d" "table"="Port_Group"
2022/02/15 18:01:28 server/cache: "caller"={"file":"cache.go","line":683} "level"=5 "msg"="updating row" "uuid"="a6114f2c-c662-4c1e-bdf4-cc83b3ffaf3d" "table"="Port_Group" "old"="&{UUID:a6114f2c-c662-4c1e-bdf4-cc83b3ffaf3d ACLs:[189fecf6-b1bf-4f67-b5aa-5d0584f7da41 13af1844-b856-4555-a9f6-161a4a3eeeea] ExternalIDs:map[name:a6953372168492035427_ingressDefaultDeny] Name:a6953372168492035427_ingressDefaultDeny Ports:[]}" "new"="&{UUID:a6114f2c-c662-4c1e-bdf4-cc83b3ffaf3d ACLs:[189fecf6-b1bf-4f67-b5aa-5d0584f7da41 13af1844-b856-4555-a9f6-161a4a3eeeea] ExternalIDs:map[name:a6953372168492035427_ingressDefaultDeny] Name:a6953372168492035427_ingressDefaultDeny Ports:[413294c1-2463-4d2f-afc6-8a04e2461721]}"
2022/02/15 18:01:28 server/cache: "caller"={"file":"cache.go","line":650} "level"=5 "msg"="processing update" "uuid"="dfefa173-7537-4d7b-a684-6843d52ddbd7" "table"="Port_Group"
2022/02/15 18:01:28 server/cache: "caller"={"file":"cache.go","line":683} "level"=5 "msg"="updating row" "uuid"="dfefa173-7537-4d7b-a684-6843d52ddbd7" "table"="Port_Group" "old"="&{UUID:dfefa173-7537-4d7b-a684-6843d52ddbd7 ACLs:[23c2636c-7ceb-41e2-85ef-388f3cef49c8 afd03668-e66d-455f-9355-f0d80504687e] ExternalIDs:map[name:a6953372168492035427_egressDefaultDeny] Name:a6953372168492035427_egressDefaultDeny Ports:[]}" "new"="&{UUID:dfefa173-7537-4d7b-a684-6843d52ddbd7 ACLs:[23c2636c-7ceb-41e2-85ef-388f3cef49c8 afd03668-e66d-455f-9355-f0d80504687e] ExternalIDs:map[name:a6953372168492035427_egressDefaultDeny] Name:a6953372168492035427_egressDefaultDeny Ports:[413294c1-2463-4d2f-afc6-8a04e2461721]}"
2022/02/15 18:01:28 server/cache: "caller"={"file":"cache.go","line":650} "level"=5 "msg"="processing update" "uuid"="a605fd16-f353-4b6c-a2c1-cea593879c24" "table"="Port_Group"
2022/02/15 18:01:28 server/cache: "caller"={"file":"cache.go","line":667} "level"=5 "msg"="inserting row" "uuid"="a605fd16-f353-4b6c-a2c1-cea593879c24" "table"="Port_Group" "model"="&{UUID:a605fd16-f353-4b6c-a2c1-cea593879c24 ACLs:[e21bd519-22c8-4112-8603-03cfa1862126 e8c7a17f-8ffd-478f-ab74-f667dc190f5a] ExternalIDs:map[name:namespace1_networkpolicy1] Name:a14195333570786048679 Ports:[413294c1-2463-4d2f-afc6-8a04e2461721]}"
2022/02/15 18:01:28 client error handling update2 rpc: connection is shut down
2022/02/15 18:01:28 cache: "caller"={"file":"cache.go","line":650} "level"=5 "msg"="processing update" "uuid"="e21bd519-22c8-4112-8603-03cfa1862126" "table"="ACL"
2022/02/15 18:01:28 cache: "caller"={"file":"cache.go","line":667} "level"=5 "msg"="inserting row" "uuid"="e21bd519-22c8-4112-8603-03cfa1862126" "table"="ACL" "model"="&{UUID:e21bd519-22c8-4112-8603-03cfa1862126 Action:allow-related Direction:to-lport ExternalIDs:map[Ingress_num:0 ipblock_cidr:false l4Match:tcp && tcp.dst==82 namespace:namespace1 policy:networkpolicy1 policy_type:Ingress] Label:0 Log:false Match:ip4 && tcp && tcp.dst==82 && outport == @a14195333570786048679 Meter:0xc001508020 Name:0xc001508030 Priority:1001 Severity:0xc001508010}"
```

Any ideas on what would cause this? Some race between Close() and Connect() + Monitor()?

Reply:
After writing a giant patch to support shutting down the server rather than the client I realized this was a red herring. update2 is the update for the _Server DB and not the NB DB. We were actually getting back the cache update for removing the PG, but libovsdb server was not removing the ACLs:
https://github.com/ovn-org/libovsdb/issues/219

I put in a workaround for now to also expect the leftover ACLs to exist.